Bug fixes for method typing

fo5for opened the pull request:
Issue: subsumption checking was not performed on propagated method definitions, allowing this to be typed:
```scala
class Test4A
    method Mth4A: anything
    method Mth4A = true
//│ Defined class Test4A
//│ Declared Test4A.Mth4A: test4A -> true
//│ Defined Test4A.Mth4A: test4A -> true
:e
class Test4B: Test4A
    method Mth4A: int
//│ Defined class Test4B
//│ Declared Test4B.Mth4A: test4B -> true
```

It now emits an error:
https://github.com/hkust-taco/mlscript/blob/1e7cee08b7a7ae7de95b52249cc1aea5da33d855/shared/src/test/diff/mlscript/BadMethods.mls#L373-L384

Side effect: the error message are improved in some cases:
https://github.com/hkust-taco/mlscript/compare/048e0a1..1e7cee0#diff-271f81eabb0cf04b1f9fe33d0620031626c2c2cb6c311ebafa625a54caf951cbL144-R150

fo5for committed:
Check subsumption for propagated method definitions

LPTK commented:
Good! Though on a related note, I'm surprised that the type of `Test4A.Mth4A` in your example is `true` instead of `anything`. Explicit method signatures should take precedence over inferred types, at least when viewed from the outside of the class (probably from the inside too). Otherwise requiring explicit signatures when overriding makes little sense, as the signature is ignored anyway.

LPTK commented:
@fo5for maybe you want to address that in this PR to kill two birds with one stone?

fo5for commented:
The way I intended for it to work is that implicit calls uses the declared signature and explicit calls uses the inferred signature, so the caller can choose which one to use. (It's not working in the current version due to a bug...)

```scala
o = Test4A {}
o.Mth4A
o.(Test4A.Mth4A)
//│ o: test4A
//│ res: anything
//│ res: true
```

I realized that this may not be possible if the class inherits from another class that declared the same method. So I'll change it to always use the declared signature.


> @fo5for maybe you want to address that in this PR to kill two birds with one stone?

Sure. I will actually kill 3 birds by making diffTests print the declared and defined signature correctly. It was printing the defined signature twice if they are both present.

fo5for committed:
Bug fixes for method typing

- Declared method signature always takes precedence over inferred
  signature
- Fix explicit type variables for method definitions
- Fix method signature printing in `DiffTests`

LPTK commented on shared/src/test/diff/mlscript/Methods.mls:
```diff
@@ -10,20 +12,20 @@ class Bar: Foo[int, bool]
 
 f.(Foo.Fun)
-//│ res: (('A | 1) -> ('B | 2) -> 'C) -> 'A -> 'B -> 'C
+//│ res: (nothing -> anything & 'a) -> 'a
```
These diffs are very fishy. I wonder what's happening here.

fo5for replied:
I reverted the precedence fix to test the other fixes independently, then applied it back manually before committing. Turns out I applied it in the wrong place 😅

LPTK commented on shared/src/test/scala/mlscript/DiffTests.scala:
```diff
@@ -253,9 +253,12 @@ class DiffTests extends org.scalatest.funsuite.AnyFunSuite {
                   // ^ it may not end up being defined if there's an error
                 output(s"Defined " + td.kind.str + " " + td.nme.name)
                 val ttd = ctx.tyDefs(td.nme.name)
+                val thisTy = typer.TypeRef(ttd, ttd.tparams.flatMap(p => 
```
Is there an easy way of avoiding too much logic duplication in `DiffTests`? It is very likely that code manipulating type provenance etc. will change in future Typer updates, and this would create a discrepancy. The logic should at least be moved and shared in a Typer method.

LPTK commented:
@fo5for Let me know when the PR is ready for review and merge.

fo5for committed:
Refactor duplicated logic
